### PR TITLE
add linked avatar

### PR DESCRIPTION
**Pull request by ewerx**

### CODE REVIEW

**Review by triyambagam — APPROVED** [review]
Continuing with my changes

**Comment by ewerx** [comment]
@triyambagam updated this branch

**Review by gabsong — APPROVED** [review]
LGTM

**Review by 6rivu — APPROVED** [review]
LGTM

File: contracts/ProfileToken.sol [hunk 1]
```diff
@@ -13,25 +13,62 @@ contract ProfileToken is
...
+        _avatarInfo.nftContract = avatarContractAddrs;
+        _avatarInfo.tokenId = avatarTokenId;
+
+        emit SetAvatar(niftyTokenId, avatarContractAddrs, avatarTokenId, tokenURI(niftyTokenId));
+
+        // console.log("avatar: %s/%s => profile %s", avatarContractAddrs, avatarTokenId, niftyTokenId);
+    }
+
+    // pass through to avatar contract
+    function tokenURI(uint256 /*tokenId*/) public view virtual override(ERC721) returns (string memory) {
```
Comment (ewerx) [inline contracts/ProfileToken.sol]: @triyambagam pointed out this only returns the last set avatar, we need to store a mapping and use the token id passed in to look up

File: contracts/ProfileToken.sol [hunk 2]
```diff
@@ -13,25 +13,62 @@ contract ProfileToken is
...
+    function setAvatar(address avatarContractAddrs, uint avatarTokenId, uint niftyTokenId) public {
+        require(msg.sender == ERC721.ownerOf(niftyTokenId), "not profile owner");
+
+        ERC721 avatarNFT = ERC721(avatarContractAddrs);
+        require(msg.sender == avatarNFT.ownerOf(avatarTokenId), "not avatar owner");
+
+        _avatarInfo.nftContract = avatarContractAddrs;
+        _avatarInfo.tokenId = avatarTokenId;
+
+        emit SetAvatar(niftyTokenId, avatarContractAddrs, avatarTokenId, tokenURI(niftyTokenId));
```
Comment (gabsong) [inline contracts/ProfileToken.sol]: What is the `tokenURI(niftyTokenId)` doing? I think at this point, the result of this is just an empty string. I may be missing something, so can you please explain?

Comment (gabsong) [reply contracts/ProfileToken.sol]: Oh never mind, I saw the overriding `tokenURI` function below.

File: contracts/ProfileToken.sol [hunk 3]
```diff
@@ -13,25 +13,62 @@ contract ProfileToken is
...
+
+        // console.log("avatar: %s/%s => profile %s", avatarContractAddrs, avatarTokenId, niftyTokenId);
+    }
+
+    // pass through to avatar contract
+    function tokenURI(uint256 /*tokenId*/) public view virtual override(ERC721) returns (string memory) {
+        //TODO: return a default profile URI if avatar is not set
+        IERC721Metadata avatarNFT = IERC721Metadata(_avatarInfo.nftContract);
```
Comment (gabsong) [inline contracts/ProfileToken.sol]: Can we add a comment that explains what this line does?

File: contracts/ProfileToken.sol [hunk 4]
```diff
@@ -52,30 +51,47 @@ contract ProfileToken is
...
+        );
     }
 
     // pass through to avatar contract
-    function tokenURI(uint256 tokenId) public view virtual override(ERC721) returns (string memory) {
-
-        require(_exists(tokenId), "Querying for a non existent token");
-
-         return _tokenURIs[tokenId];
+    function tokenURI(uint256 tokenId)
```
Comment (gabsong) [inline contracts/ProfileToken.sol]: I think setting the parameter to `niftyTokenId` would make the code more clear.

File: contracts/ProfileToken.sol [hunk 5]
```diff
@@ -52,30 +51,47 @@ contract ProfileToken is
...
+        public
+        view
+        virtual
+        override(ERC721)
+        returns (string memory)
+    {
+        require(_exists(tokenId), "non-existent token");
+        IERC721Metadata avatarNFT = IERC721Metadata(_avatars[tokenId].nftContract);
```
Comment (gabsong) [inline contracts/ProfileToken.sol]: @ewerx can you explain this line? Sorry if this is supposed to be something very obvious, but I'm not understanding it!

File: contracts/ProfileToken.sol (hunk 4 above)
Comment (ewerx) [reply contracts/ProfileToken.sol]: this is an override can we rename the params?

Comment (ewerx) [reply contracts/ProfileToken.sol]: seems ok, updated

File: contracts/ProfileToken.sol (hunk 5 above)
Comment (ewerx) [reply contracts/ProfileToken.sol]: added comment

**Pull request merged**